### ISSUE TITLE
Missed `manual_filter_map` with `matches!`

### ISSUE THREAD

**Issue by dodomorandi** [rust-lang/rust-clippy#8010]
Lint name:
`manual_filter_map`

I tried this code:

```rust
iter.filter(|x| matches!(x, Enum::A(_)))
    .map(|x| if let Enum::A(s) = x { s } else { unreachable!() })
```
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=8a10a5c2d8647c8b3f93e8f64448c961)

I expected to see this happen: `manual_filter_map` lint should be triggered.

Instead, this happened: everything's fine for clippy.

### Meta

**Rust version (`rustc -Vv`):**

```
rustc 1.46.0-nightly (f455e46ea 2020-06-20)
binary: rustc
commit-hash: f455e46eae1a227d735091091144601b467e1565
commit-date: 2020-06-20
host: x86_64-unknown-linux-gnu
release: 1.46.0-nightly
LLVM version: 10.0
```

### Additional information
There are many ways to express the _filter_ part:
* `matches!`
* `if let /* ... */ { true } else { false }`
* `match /* ... */ => true, _ => false `
Similarly, for the _map_ part it is possible to use:
* `if let /* ... */ { inner } else { /* diverge */ }`
* `match /* ... */ => inner, _ => /* diverge */ `
* (soon) `let inner = /* ... */ else { /* diverge */ }`

Moreover, I am not considering eventual helper functions implemented for the enum like `fn is_a(&self) -> bool` or `fn unwrap_a(self) -> A`.
As you can see, the number of cases is combinatorial. Therefore it is reasonable to match the patterns for the _filter_ and the _map_ parts independently, and use the results to perform the final detection. On the other hand, I am worried that trying to _generalize_ the detection of `manual_filter_map` in this way could lead to false positives.
